File: docs/specs/episode-prompt-load-evaluation/simple-rules/reproduction/analyze.py
"""저장한 출력의 조건 가리기와 관찰 집계. 모델을 호출하지 않는다."""

import hashlib
import json
import statistics
import sys
from pathlib import Path


directory = Path(sys.argv[2])


def read(name):
    return json.loads((directory / name).read_text())


def write(name, value):
    (directory / name).write_text(json.dumps(value, ensure_ascii=False, indent=2) + "\n")


rows = [json.loads(line) for line in (directory / "responses.jsonl").read_text().splitlines()]
manifest = read("manifest.json")
expected_ids = {job["id"] for job in manifest["jobs"]}
assert len(rows) == len(expected_ids)
assert {row["id"] for row in rows} == expected_ids
manifest_sha = hashlib.sha256((directory / "manifest.json").read_bytes()).hexdigest()
assert all(row["manifestSha256"] == manifest_sha for row in rows)
fixtures = {fixture["id"]: fixture for fixture in manifest["scenarios"]}

if sys.argv[1] == "mask":
    assert not (directory / "review.json").exists(), "관찰한 뒤 검토 번호를 다시 만들지 않는다."
    ordered = sorted(rows, key=lambda row: hashlib.sha256(("simple-review:" + row["id"]).encode()).hexdigest())
    masked = [{"reviewId": f"V{i + 1:03d}", "scenarioId": row["scenarioId"],
               "scene": row.get("answer", {}).get("scene"), "error": row.get("error")}
              for i, row in enumerate(ordered)]
    write("masked-scenes.json", masked)
    write("review-map.json", {item["reviewId"]: row["id"] for item, row in zip(masked, ordered)})
    (directory / "masked-dialogue.txt").write_text("\n".join(
        item["reviewId"] + " " + item["scenarioId"] + " | " + " | ".join(
            line["speaker"] + ": " + line["text"] for line in (item["scene"] or {}).get("dialogue", []))
        for item in masked) + "\n")
    (directory / "masked-endings.txt").write_text("\n".join(
        item["reviewId"] + " " + item["scenarioId"] + " | " + json.dumps(
            (item["scene"] or {}).get("ending"), ensure_ascii=False) for item in masked) + "\n")
    print(f"Masked {len(masked)} outputs")
elif sys.argv[1] == "summarize":
    review = read("review.json")
    mapping = read("review-map.json")
    assert set(review["reviewedIds"]) == set(mapping)
    notes = {mapping[note["reviewId"]]: note for note in review["observations"]}
    assert len(notes) == len(review["observations"])
    categories = ["wrong-person", "invented-schedule", "ambiguous-recipient", "missing-answer", "other-fact"]

    def aggregate(subset):
        answers = [row["answer"] for row in subset if "answer" in row]
        def issues(category):
            return [row["id"] for row in subset if category in notes.get(row["id"], {}).get("categories", [])]
        return {
            "n": len(subset),
            "requestErrors": [row["id"] for row in subset if row.get("error")],
            "mechanicalFailures": [row["id"] for row in subset if row.get("error") or row["problems"]],
            "issues": {category: issues(category) for category in categories},
            "scheduleErrors": [row["id"] for row in subset if any(category in notes.get(row["id"], {}).get("categories", []) for category in categories[:2])],
            "requiredAnswers": sum(bool(fixtures[row["scenarioId"]]["requiredMention"]) for row in subset),
            "correctEnding": sum("answer" in row and (row["answer"]["scene"]["ending"] or {}).get("kind") == fixtures[row["scenarioId"]]["expectedEnding"] for row in subset),
            "oneOrTwoUtterances": sum(1 <= len(answer["scene"]["dialogue"]) <= 2 for answer in answers),
            "costUsd": sum(answer["gatewayCostUsd"] for answer in answers if answer["gatewayCostUsd"] is not None),
            "missingCost": sum(answer["gatewayCostUsd"] is None for answer in answers),
            "medianFirstDialogueMs": statistics.median(answer["deltas"][0]["elapsedMs"] for answer in answers if answer["deltas"]) if answers else None,
            "medianElapsedMs": statistics.median(answer["elapsedMs"] for answer in answers) if answers else None,
        }

    summary = {"calls": len(rows), "manifestSha256": manifest_sha,
               "models": sorted({row.get("answer", {}).get("model", "error") for row in rows}),
               "variants": {variant: aggregate([row for row in rows if row["variant"] == variant]) for variant in ["before", "after"]},
               "byScenario": {scenario_id: {variant: aggregate([row for row in rows if row["scenarioId"] == scenario_id and row["variant"] == variant]) for variant in ["before", "after"]} for scenario_id in fixtures}}
    write("summary.json", summary)
    print(json.dumps({key: value for key, value in summary.items() if key != "byScenario"}, ensure_ascii=False, indent=2))
else:
    raise ValueError("Usage: analyze.py <mask|summarize> <data-directory>")

File: docs/specs/episode-prompt-load-evaluation/role-check/reproduction/analyze.ts
import { readFileSync, writeFileSync } from "node:fs";
import { createHash } from "node:crypto";
import { resolve } from "node:path";
const dir = resolve(import.meta.dir, "data");
const read = (name: string) => JSON.parse(readFileSync(resolve(dir, name), "utf8"));
const write = (name: string, value: unknown) => writeFileSync(resolve(dir, name), JSON.stringify(value, null, 2) + "\n");
const rows = readFileSync(resolve(dir, "responses.jsonl"), "utf8").trim().split("\n").map(s => JSON.parse(s));
const scenes = rows.filter(r => r.arm !== "facts");
const m = read("manifest.json");
const sha = (s: string) => createHash("sha256").update(s).digest("hex");
if (process.argv[2] === "mask") {
  if (rows.length !== 160 || scenes.length !== 120) throw new Error("Incomplete run");
  const ordered = [...scenes].sort((a, b) => sha(`roles-blind:${a.id}`).localeCompare(sha(`roles-blind:${b.id}`)));
  const blind = ordered.map((r, i) => ({reviewId: `V${String(i + 1).padStart(3, "0")}`, fixtureId: r.fixtureId, dialogue: r.scene?.dialogue ?? [], error: r.error ?? null}));
  write("blind-dialogue.json", blind);
  write("blind-map.json", Object.fromEntries(blind.map((b, i) => [b.reviewId, ordered[i].id])));
  writeFileSync(resolve(dir, "blind-dialogue.txt"), blind.map(b => `${b.reviewId} ${b.fixtureId} | ${b.dialogue.map((d: any) => `${d.speaker}: ${d.text}`).join(" | ")}`).join("\n") + "\n");
  console.log(`Masked ${blind.length} scene outputs`);
} else if (process.argv[2] === "summarize") {
  if (rows.length !== m.calls || new Set(rows.map(r => r.id)).size !== m.calls) throw new Error("Missing or duplicate call");
  const notes = read("review.json");
  const map = read("blind-map.json");
  const notesByJob = Object.fromEntries(notes.observations.map((n: any) => [map[n.reviewId], n]));
  const facts = rows.filter(r => r.arm === "facts");
  const index = Object.fromEntries(rows.map(r => [r.id, r]));
  const canon = (xs: any[]) => xs.map(x => `${x.person}:${x.event}`).sort().join("|");
  const factChecks = facts.map(r => ({id: r.id, expected: m.cases.find((f: any) => f.id === r.fixtureId).expectedAppointments, actual: r.facts?.appointments ?? null, match: r.facts && canon(r.facts.appointments) === canon(m.cases.find((f: any) => f.id === r.fixtureId).expectedAppointments)}));
  const checks = scenes.map(r => {
    const f = m.cases.find((f: any) => f.id === r.fixtureId);
    const note = notesByJob[r.id] ?? null;
    return {id: r.id, fixtureId: r.fixtureId, arm: r.arm, schemaValid: r.valid === true, firstDialogueMs: r.firstDialogueMs, elapsedMs: r.elapsedMs, utterances: r.scene?.dialogue.length ?? 0, endingMatches: r.scene?.ending?.kind === f.expectedEnding, note, requiredMention: f.requiredMention, factSourceMatches: r.factSourceId ? index[r.factSourceId].factsMatch : null};
  });
  const total = (rs: any[], fn: (r: any) => number) => rs.reduce((n, r) => n + fn(r), 0);
  const median = (xs: number[]) => [...xs].sort((a, b) => a - b)[Math.ceil(xs.length / 2) - 1];
  const arms = ["A", "R", "S"].map(arm => {
    const rs = scenes.filter(r => r.arm === arm);
    const cs = checks.filter(c => c.arm === arm);
    const issues = (rs: any[], code: string) => rs.filter(r => r.note?.categories.includes(code)).length;
    return {arm, n: rs.length, schemaValid: cs.filter(c => c.schemaValid).length, wrongOwner: issues(cs, "wrong-owner"), unsupportedSchedule: issues(cs, "unsupported-schedule"), ambiguous: issues(cs, "ambiguous-target"), missingRequired: issues(cs, "missing-required"), requiredResponseCount: cs.filter(c => c.requiredMention.length).length, correctEnding: cs.filter(c => c.endingMatches).length, oneOrTwoUtterances: cs.filter(c => c.utterances >= 1 && c.utterances <= 2).length, sceneCostUsd: total(rs, r => Number(r.gateway?.cost ?? 0)), totalCostWithExtractionUsd: total(rs, r => Number(r.gateway?.cost ?? 0) + (r.factSourceId ? Number(index[r.factSourceId].gateway?.cost ?? 0) : 0)), medianSceneFirstDialogueMs: median(rs.map(r => r.firstDialogueMs)), medianSceneElapsedMs: median(rs.map(r => r.elapsedMs)), estimatedSerialFirstDialogueMs: median(rs.map(r => r.firstDialogueMs + (r.factSourceId ? index[r.factSourceId].elapsedMs : 0))), estimatedSerialElapsedMs: median(rs.map(r => r.elapsedMs + (r.factSourceId ? index[r.factSourceId].elapsedMs : 0))), byFixture: m.cases.map((f: any) => {const subset = cs.filter(c => c.fixtureId === f.id); return {fixtureId: f.id, wrongOwner: issues(subset, "wrong-owner"), unsupportedSchedule: issues(subset, "unsupported-schedule"), ambiguous: issues(subset, "ambiguous-target"), missingRequired: issues(subset, "missing-required"), correctEnding: subset.filter(c => c.endingMatches).length};})};
  });
  const summary = {calls: rows.length, requestErrors: rows.filter(r => r.error).map(r => ({id: r.id, error: r.error})), models: [...new Set(rows.map(r => r.response?.modelId))], providers: [...new Set(rows.map(r => r.gateway?.routing?.finalProvider))], costUsd: total(rows, r => Number(r.gateway?.cost ?? 0)), factExtraction: {n: facts.length, exactMatches: factChecks.filter(f => f.match).length, costUsd: total(facts, r => Number(r.gateway?.cost ?? 0)), medianMs: median(facts.map(f => f.elapsedMs))}, arms};
  write("fact-checks.json", factChecks);
  write("checks.json", checks);
  write("summary.json", summary);
  console.log(JSON.stringify(summary, null, 2));
} else {throw new Error("Use mask or summarize");}
